pkg/client/root_service.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2019-2020 vChain, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"github.com/codenotary/immudb/pkg/api/schema"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"google.golang.org/grpc"
+	"io/ioutil"
+)
+
+const ROOT_FN = ".root"
+
+type RootService interface {
+	GetRoot(ctx context.Context) (*schema.Root, error)
+}
+
+type rootservice struct {
+	client schema.ImmuServiceClient
+}
+
+func NewRootService(immuC schema.ImmuServiceClient) RootService {
+	return &rootservice{immuC}
+}
+
+func (r *rootservice) GetRoot(ctx context.Context) (*schema.Root, error) {
+	if root, err := GetCachedRoot(); err == nil {
+		return root, nil
+	}
+	var protoReq empty.Empty
+	var metadata runtime.ServerMetadata
+	if root, err := r.client.CurrentRoot(ctx, &protoReq, grpc.Header(&metadata.HeaderMD), grpc.Trailer(&metadata.TrailerMD)); err != nil {
+		return nil, err
+	} else {
+		if err := SetRoot(root); err != nil {
+			return nil, err
+		}
+		return root, nil
+	}
+}
+
+func SetRoot(root *schema.Root) error {
+	var buf bytes.Buffer
+	raw, err := root.XXX_Marshal(buf.Bytes(), true)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(ROOT_FN, raw, 0644)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+
+func GetCachedRoot() (*schema.Root, error) {
+	root := new(schema.Root)
+	buf, err := ioutil.ReadFile(ROOT_FN)
+	if err == nil {
+		if err = root.XXX_Unmarshal(buf); err != nil {
```

Comment:
We should use the `github.com/golang/protobuf/proto` instead of accessing `XXX_` functions:
```suggestion
		if err = proto.Unmarshal(buf, root); err != nil {
```